### DISCUSSION TITLE
Comparison between Different Types (operator ==)

### DISCUSSION THREAD

**Discussion by stellarpower** [nushell/nushell#10891] [General]
Before I go further into some motivation - is there a design choice that we can compare values of incompatible types?

```nu
~> "true" == true                                                                                                                                    
false
~> "3" == 4   
false
~> [1 2 3] == "world" 
false
~> "three" == null                                                                                              
false

```

I feel these should all be type errors, rather than returning false. I hit a nasty bug due to a typo, and I feel if there's no way any comparison between those two types could ever be true, then it should probably be a type error.

An example of a comparison that could be true, but might not be, would be comparing floats and integers. Implicitly promoting the integer and comparing equal types is a nice way to do it, but even if it's strictly an int-float or float-int overload of an operator, it's at least possible to mix them that way.

However comparing a string to a float or null to anything but maybe itself or a boolean I think makes no sense in nushell - so surely, shouldn't we disallow it? JavaScript went down this path a long time ago and gained a lot of notoriety, so I would be interested if there's a compelling reason why this behaviour exists currently and if we need it going forwards.

For lists or columns or other iterables, the question of vectorisation comes in - but then, using a separate operator for the syntax is probably much much safer.

Cheers

**Comment by amtoine** [comment]
i would say comparing with `null` makes sense.
the rest, yeah, maybe they should give an error :thinking:

**Reply by stellarpower** [reply]
Are types nullable in this context? If so, then I agree. But I thought it was like some other languages where `string?` is  more like `string | nothing` rather than a pointer to a string. In which case comparing a string (literal) to null shouldn't ever work, but comparing a nullable string to null would totally make sense. But then you'd need to watch what happens when you compare a nullable string to a string. Do we return false or is that a runtime error do dereference null?

**Comment by Tiggax** [comment]
I'm thinking that for lists maybe like @stellarpower  said in the [discord channel](https://discord.com/channels/601130461678272522/614593951969574961/1168599429167726602) the possibility of equality translating to objects in a list.
This would make the language more similar to R. 
The question is if this would feel intuitive for people using nushell, but other than comparing lists (The current implementation needs the items in a list to be exact, and in the same order so `[1 2] == [2 1]` is `false` ), this addition wouldn't feel too bad in my opinion.

**Reply by stellarpower** [reply]
Just to confirm if you're referring to something like `[1 2 3] == 1` resulting in `[true false false]`, then I think it's a useful idea, but overall it feels safer if a different syntax is used. I agree it's intuitive though, and at least if you get a list back you can't co-erce that to a boolean without `all` or `any`

**Reply by Tiggax** [reply]
Yes, this is what happens in R:  every operation on vectors (lists) is translated to the elements in that list (unless the function operates on list).
This can currently be achieved by `[1 2 3] | each {|i| $i == 1}` 
If you are looking for `any`  implementation, currently you could just do:`1 in [1 2 3]`

**Reply by fdncred** [reply]
we also have an `any` command with examples

**Reply by stellarpower** [reply]
Julia uses a dot for this use AFAIK, and I think that makes sense. `[1 2 3] .== 1]`

When I first stumbled across this, I thought nushell was vectorising across the list as you described in fact. But it seems to be more of a core feature across types.
